features/294-login-without-specifying-login-id-key/api.md

```diff
@@ -0,0 +1,374 @@
+4. [skygear-server] Update signup and login handler to handle the `loginID` concept.
+5. [skygear-server] Include `LoginID-Source` in response header.
+6. [skygear-server] Implement `requestVerification(LoginIDSource: String, type: "email"|"phone")`
+
+### SSO auto-link
+
+For SSO auto-link user feature, have following possible solutions:
+
+1. assume `loginID` is email.
+2. provide an extra auth-hook, which request to return `userID` by email.
+3. Check if there is email information in the metadata:
+   1. if there is a `sso_matching_emails`, all emails in the array of value will be matched against the SSO login.
+   2. if ther is a `email`, that email will be matched against the SSO login.
+
+Solution 3 may have uniqueness issue, i.e. two user could have identical email or overlapping email in `sso_matching_emails`.
+
+## Proposal 3
+
+Adjust current `LOGIN_ID_METADATA_KEYS` definition, rename it `LOGIN_IDS_KEY_WHITELIST` and allow one level string `loginID` only. `LOGIN_IDS_KEY_WHITELIST` default to empty, which allows to use any string as the `loginIDSouce`.
+
+```
+LOGIN_IDS_KEY_WHITELIST = []
+==> allows any string as loginIDSouce
+
+LOGIN_IDS_KEY_WHITELIST = ["username", "email"]
+==> allows "username" and "email" as loginIDSouce
+```
+
+And `loginIDData` must be following form:
+
+```
+{
+  <loginSource: String>: <loginId: String>,
+  <loginSource: String>: <loginId: String>
+}
+```
+
+For example, for `LOGIN_IDS_KEY_WHITELIST = []`,
+
+```
+signup({
+  username: 'example',
+  email: 'example@example.com'
+}, 'password')
+```
+
+which creates two `loginID`s:
+
+| login_id_source:String | login_id:String (`UNIQUE`) |
+| --- | --- |
+| username | example |
+| email | example@example.com |
+
+User can login by `username` **OR** `email`, but not `username` **AND** `email`.
+
+```
+login('example', 'password'); ==> OK
+login('example@example', 'password'); ==> OK
+login({
+  username: 'example',
+  email: 'example@example.com'
+}, 'password') ==> won't support this.
+```
+
+Since `loginID` is defined with `UNIQUE` constraint, which means another user **can't** signup as
+
+```
+signup({
+  username: 'example@example.com',
+  email: 'example'
+}, 'password') ==> error: Duplicated loginID
+```
+
+For app wants to use compound keys as `loginID`, we suggest it encodes compound keys to a JSON string.
+
+```
+LOGIN_IDS_KEY_WHITELIST = ["email", "role_phone"]
+
+signup({
+  email: 'example@example.com',
+  role_phone: '["admin","phone"]'
+}, 'password') ==> OK
+```
+
+which creates two `loginID`s:
+
+| login_id_source:String | login_id:String (`UNIQUE`) |
+| --- | --- |
+| email | example@example.com |
+| role_phone | ["admin","phone"] |
+
+It's developer's responsibility to ensure the encoded string uniqueness ('["admin","phone"]' and '["phone","admin"]' are considered as two different `loginID`). 
+
+If `LOGIN_IDS_KEY_WHITELIST ` contains some string (not empty), a user can't signup with a `loginIDSouce` not in the list.
+
```

Comment:
Typo

Reply:
I can't find the typo. 

If you mean `loginIDSouce` is the typo in the quoted section, `loginSouce` is typo actually. Anyway, I remove the term and call it `loginIDKey` in the API.

Reply:
:s/souce/source/g

Reply:
Sorry for this.